vehicle/remote_control_process.py

```diff
@@ -246,1086 +306,939 @@ def close_motor_socket(self):
+        fatal_error = False
+        zero_output = False
+
+        if self.motor_state != _STATE_ENABLED:
+            error_messages.append("Motor error so zeroing out autonomy commands.")
+            zero_output = True
+            self.control_state = CONTROL_MOTOR_ERROR
+            self.resume_motion_timer = time.time()
+
+        if self.voltage_average < _VOLTAGE_CUTOFF:
+            fatal_error = True
+            error_messages.append("Voltage low so zeroing out autonomy commands.")
+            zero_output = True
+            self.control_state = CONTROL_LOW_VOLTAGE
+            self.resume_motion_timer = time.time()
+
+        if not self.gps.is_dual_fix():
+            error_messages.append("No GPS fix so zeroing out autonomy commands.")
+            zero_output = True
+            self.control_state = CONTROL_GPS_STARTUP
+            self.resume_motion_timer = time.time()
+        elif abs(absolute_path_distance) > self.nav_path.maximum_allowed_distance_meters:
+            # Distance from path exceeds allowed limit.
+            zero_output = True
+            self.resume_motion_timer = time.time()
+            self.control_state = CONTROL_AUTONOMY_ERROR_DISTANCE
+            error_messages.append(
+                "GPS distance {} meters too far from path so zeroing out autonomy commands."
+                .format(abs(absolute_path_distance)))
+        elif abs(gps_path_angle_error) > self.nav_path.maximum_allowed_angle_error_degrees:
+            zero_output = True
+            self.resume_motion_timer = time.time()
+            self.control_state = CONTROL_AUTONOMY_ERROR_ANGLE
+            error_messages.append(
+                "GPS path angle {} exceeds allowed limit {} so zeroing out autonomy commands."
+                .format(
+                    abs(gps_path_angle_error),
+                    self.nav_path.maximum_allowed_angle_error_degrees))
+        elif self.gps.last_sample().rtk_age > _ALLOWED_RTK_AGE_SEC:
+            zero_output = True
+            self.resume_motion_timer = time.time()
+            self.control_state = CONTROL_AUTONOMY_ERROR_RTK_AGE
+            error_messages.append("RTK base station data too old so zeroing out autonomy commands.")
+        elif time.time() - self.gps.last_sample().time_stamp > _ALLOWED_SOLUTION_AGE_SEC:
+            zero_output = True
+            self.resume_motion_timer = time.time()
+            self.control_state = CONTROL_AUTONOMY_ERROR_SOLUTION_AGE
+            error_messages.append("RTK solution too old so zeroing out autonomy commands.")
+
+        if time.time() - self.robot_object.last_server_communication_stamp > SERVER_COMMUNICATION_DELAY_LIMIT_SEC:
+            zero_output = True
+            self.resume_motion_timer = time.time()
+            error_messages.append(
+                f"Server communication error so zeroing out autonomy commands. "
+                f"Last stamp age {time.time() - self.robot_object.last_server_communication_stamp} "
```

Comment:
Note the use of f-string in certain cases. Similar to string interpolation in Javascript, this embeds the expression right within the format string.